File: app/src/main/java/posidon/launcher/wall/WallActivity.java
package posidon.launcher.wall;

import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.palette.graphics.Palette;
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat;

import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.snackbar.Snackbar;

import java.io.File;
import java.io.FileOutputStream;
import java.lang.ref.WeakReference;
import java.util.Objects;

import posidon.launcher.R;
import posidon.launcher.tools.Loader;
import posidon.launcher.tools.Tools;

public class WallActivity extends AppCompatActivity {

    public static Bitmap img;
    private ImageView loading;
    private int index;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Tools.applyFontSetting(this);
        setContentView(R.layout.wall_preview);
        loading = findViewById(R.id.loading);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
        if (img.getHeight() / img.getWidth() < getResources().getDisplayMetrics().heightPixels / getResources().getDisplayMetrics().widthPixels)
            img = Tools.centerCropWallpaper(WallActivity.this, img);
        ((ImageView)findViewById(R.id.theimg)).setImageBitmap(img);
        final Bundle extras = getIntent().getExtras();
        if (extras == null) {
            loading.setVisibility(View.GONE);
            findViewById(R.id.downloadbtn).setVisibility(View.GONE);
        } else {
            Tools.animate(loading.getDrawable());
            index = extras.getInt("index");
            new Loader.bitmap(Gallery.walls.get(index).url, new Loader.bitmap.Listener() {
                @Override
                public void onFinished(Bitmap img) {
                    if (img.getHeight() / img.getWidth() < getResources().getDisplayMetrics().heightPixels / getResources().getDisplayMetrics().widthPixels)
                        img = Tools.centerCropWallpaper(WallActivity.this, img);
                    //else img = Bitmap.createBitmap(img, 0, 0, getResources().getDisplayMetrics().widthPixels, getResources().getDisplayMetrics().widthPixels/img.getWidth()*img.getHeight());
                    ((ImageView)findViewById(R.id.theimg)).setImageBitmap(img);
                    WallActivity.img = img;
                    if (loading.getDrawable() instanceof AnimatedVectorDrawable) {
                        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) ((AnimatedVectorDrawable)loading.getDrawable()).clearAnimationCallbacks();
                        else ((AnimatedVectorDrawable)loading.getDrawable()).stop();
                    } else ((AnimatedVectorDrawableCompat) loading.getDrawable()).clearAnimationCallbacks();
                    loading.setVisibility(View.GONE);
                }
            }).execute();
            findViewById(R.id.downloadbtn).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) { saveBitmap(img, Gallery.walls.get(index).name); }
            });
            findViewById(R.id.downloadbtn).setBackground(btnBG());
            try {
                ((TextView) findViewById(R.id.nametxt)).setText(Gallery.walls.get(index).name);
                ((TextView)findViewById(R.id.authortxt)).setText(Gallery.walls.get(index).author);
            } catch (Exception ignore) {}
        }
        int bottompadding = Tools.navbarHeight;
        if (bottompadding == 0) bottompadding = (int)(20 * getResources().getDisplayMetrics().density);
        findViewById(R.id.bottomstuff).setPadding(0,0, 0, bottompadding);
        findViewById(R.id.applybtn).setBackground(btnBG());
        findViewById(R.id.applybtn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                findViewById(R.id.bottomstuff).animate().alpha(0);
                final BottomSheetDialog dialog = new BottomSheetDialog(WallActivity.this, R.style.bottomsheet);
                dialog.setContentView(R.layout.wall_apply_dialog);
                Objects.requireNonNull(dialog.getWindow()).findViewById(R.id.design_bottom_sheet).setBackgroundColor(0x0);
                dialog.findViewById(R.id.home).setBackground(dialogBtnBG());
                dialog.findViewById(R.id.home).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) { new SetWall(img, new WeakReference<Context>(WallActivity.this), 0).execute(); dialog.dismiss(); }
                });
                dialog.findViewById(R.id.lock).setBackground(dialogBtnBG());
                dialog.findViewById(R.id.lock).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) { new SetWall(img, new WeakReference<Context>(WallActivity.this), 1).execute(); dialog.dismiss(); }
                });
                dialog.findViewById(R.id.both).setBackground(dialogBtnBG());
                dialog.findViewById(R.id.both).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) { new SetWall(img, new WeakReference<Context>(WallActivity.this), 2).execute(); dialog.dismiss(); }
                });
                dialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                        findViewById(R.id.bottomstuff).animate().alpha(1);
                    }
                });
                dialog.show();
            }
        });
        System.gc();
    }

    @Override
    protected void onPause() {
        overridePendingTransition(R.anim.slideup, R.anim.slidedown);
        super.onPause();
    }

    private ShapeDrawable btnBG() {
        float r = 24 * getResources().getDisplayMetrics().density;
        ShapeDrawable out = new ShapeDrawable(new RoundRectShape(new float[]{r, r, r, r, r, r, r, r}, null, null));
        out.getPaint().setColor(Palette.from(img).generate().getVibrantColor(0xff252627));
        return out;
    }

    private ShapeDrawable dialogBtnBG() {
        float r = 24 * getResources().getDisplayMetrics().density;
        ShapeDrawable out = new ShapeDrawable(new RoundRectShape(new float[]{r, r, r, r, r, r, r, r}, null, null));
        out.getPaint().setColor(Palette.from(img).generate().getDarkMutedColor(Palette.from(img).generate().getDominantColor(0xff111213)));
        return out;
    }

    private ShapeDrawable dialogBG() {
        float r = 40 * getResources().getDisplayMetrics().density;
        ShapeDrawable out = new ShapeDrawable(new RoundRectShape(new float[]{r, r, r, r, r, r, r, r}, null, null));
        out.getPaint().setColor(Palette.from(img).generate().getDarkMutedColor(Palette.from(img).generate().getDominantColor(0xff111213)));
        return out;
    }

    private void saveBitmap(Bitmap bitmap, String name) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);
        } else {
            File direct = (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) ? getExternalFilesDir("Wallpapers") : new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES) + "/Wallpapers");
            if (direct != null && (direct.exists() || direct.mkdirs())) {
                File file = new File(direct,name.replace(' ', '_') + ".png");
                if (file.exists()) Snackbar.make(findViewById(R.id.nametxt), "Already Saved: " + file.getAbsolutePath(), Snackbar.LENGTH_SHORT).show();
                else try {
                    FileOutputStream out = new FileOutputStream(file);
                    bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
                    out.flush();
                    out.close();
                    Snackbar.make(findViewById(R.id.nametxt), "Saved: " + file.getAbsolutePath(), Snackbar.LENGTH_SHORT).show();
                } catch (Exception e) { e.printStackTrace(); }
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == 0 && grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) saveBitmap(img, Gallery.walls.get(index).name);
    }
}

File: app/src/main/java/posidon/launcher/wall/WallActivity.kt
package posidon.launcher.wall

import android.Manifest
import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.RoundRectShape
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.palette.graphics.Palette
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.snackbar.Snackbar
import posidon.android.conveniencelib.Device
import posidon.android.conveniencelib.Graphics
import posidon.android.conveniencelib.dp
import posidon.android.conveniencelib.toBitmap
import posidon.launcher.R
import posidon.launcher.tools.*
import posidon.launcher.tools.theme.Wallpaper
import posidon.launcher.tools.theme.applyFontSetting
import java.io.File
import java.io.FileOutputStream

class WallActivity : AppCompatActivity() {

    private var loading: ImageView? = null
    private var index = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        applyFontSetting()
        setContentView(R.layout.wall_preview)
        loading = findViewById(R.id.loading)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) window.setDecorFitsSystemWindows(false)
        else window.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
        var image = img
        if (image != null && image.height / image.width < Device.screenHeight(this) / Device.screenWidth(this)) {
            image = Wallpaper.centerCropWallpaper(this, image)
        }
        findViewById<ImageView>(R.id.theimg).setImageBitmap(image)
        val extras = intent.extras
        if (extras == null) {
            loading!!.visibility = View.GONE
            findViewById<View>(R.id.downloadbtn).visibility = View.GONE
        } else {
            Graphics.tryAnimate(this, loading!!.drawable)
            index = extras.getInt("index")

            fun onImgLoaded() {
                Graphics.clearAnimation(loading!!.drawable)
                loading!!.visibility = View.GONE
                findViewById<View>(R.id.applybtn).setOnClickListener {
                    findViewById<View>(R.id.bottomstuff).animate().alpha(0f)
                    val dialog = BottomSheetDialog(this@WallActivity, R.style.bottomsheet)
                    dialog.setContentView(R.layout.wall_apply_dialog)
                    dialog.window!!.findViewById<View>(R.id.design_bottom_sheet).setBackgroundColor(0x0)
                    dialog.findViewById<View>(R.id.home)!!.background = dialogBtnBG()
                    dialog.findViewById<View>(R.id.home)!!.setOnClickListener {
                        Wallpaper.setWallpaper(image!!, 0)
                        dialog.dismiss()
                    }
                    dialog.findViewById<View>(R.id.lock)!!.background = dialogBtnBG()
                    dialog.findViewById<View>(R.id.lock)!!.setOnClickListener {
                        Wallpaper.setWallpaper(image!!, 1)
                        dialog.dismiss()
                    }
                    dialog.findViewById<View>(R.id.both)!!.background = dialogBtnBG()
                    dialog.findViewById<View>(R.id.both)!!.setOnClickListener {
                        Wallpaper.setWallpaper(image!!, 2)
                        dialog.dismiss()
                    }
                    dialog.setOnDismissListener { findViewById<View>(R.id.bottomstuff).animate().alpha(1f) }
                    dialog.show()
                }
            }

            if (Gallery.walls[index].type == Wall.Type.SVG) {
                val url = Gallery.REPO + Gallery.IMG_PATH + Gallery.walls[index].url!! + "/img.svg"
                ImageLoader.loadNullableSvg(url) { runOnUiThread {
                    onImgLoaded()
                    if (it != null) {
                        val displayWidth = Device.screenWidth(this)
                        val displayHeight = Device.screenHeight(this)
                        val width: Int
                        val height: Int
                        if (it.intrinsicHeight / it.intrinsicWidth.toFloat() < displayHeight / displayWidth.toFloat()) {
                            width = displayHeight * it.intrinsicWidth / it.intrinsicHeight
                            height = displayHeight
                        } else {
                            width = displayWidth
                            height = displayWidth * it.intrinsicHeight / it.intrinsicWidth
                        }
                        it.toBitmap(width, height).let {
                            image = it
                            if (it.height / it.width < displayHeight / displayWidth)
                                image = Wallpaper.centerCropWallpaper(this, it)
                        }
                        findViewById<ImageView>(R.id.theimg).setImageBitmap(image)
                    }
                }}
            } else {
                val url = Gallery.REPO + Gallery.IMG_PATH + Gallery.walls[index].url!! + "/img.png"
                ImageLoader.loadNullableBitmap(url) { runOnUiThread {
                    onImgLoaded()
                    if (it != null) {
                        image = it
                        val displayWidth = Device.screenWidth(this)
                        val displayHeight = Device.screenHeight(this)
                        if (it.height / it.width < displayHeight / displayWidth) image = Wallpaper.centerCropWallpaper(this, it)
                        findViewById<ImageView>(R.id.theimg).setImageBitmap(image)
                    }
                }}
            }
            findViewById<View>(R.id.downloadbtn).setOnClickListener { saveBitmap(image!!, Gallery.walls[index].name!!) }
            findViewById<View>(R.id.downloadbtn).background = btnBG()
            try {
                findViewById<TextView>(R.id.nametxt).text = Gallery.walls[index].name
                findViewById<TextView>(R.id.authortxt).text = Gallery.walls[index].author
            } catch (ignore: Exception) {}
        }
        var bottompadding = Tools.navbarHeight
        if (bottompadding == 0) bottompadding = dp(20).toInt()
        findViewById<View>(R.id.bottomstuff).setPadding(0, 0, 0, bottompadding)
        findViewById<View>(R.id.applybtn).background = btnBG()
        System.gc()
    }

    override fun onPause() {
        overridePendingTransition(R.anim.slideup, R.anim.slidedown)
        super.onPause()
    }

    private fun btnBG(): ShapeDrawable {
        val r = dp(24)
        val out = ShapeDrawable(RoundRectShape(floatArrayOf(r, r, r, r, r, r, r, r), null, null))
        out.paint.color = Palette.from(img!!).generate().getVibrantColor(-0xdad9d9)
        return out
    }

    private fun dialogBtnBG(): ShapeDrawable {
        val r = dp(24)
        val out = ShapeDrawable(RoundRectShape(floatArrayOf(r, r, r, r, r, r, r, r), null, null))
        out.paint.color = Palette.from(img!!).generate().getDarkMutedColor(Palette.from(img!!).generate().getDominantColor(-0xeeeded))
        return out
    }

    private fun saveBitmap(bitmap: Bitmap, name: String) {
        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 0)
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val values = ContentValues().apply {
                put(MediaStore.Images.Media.DISPLAY_NAME, name)
                put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg")
                put(MediaStore.Images.Media.RELATIVE_PATH, "Pictures/posidon walls/")
                put(MediaStore.Images.Media.IS_PENDING, 1)
            }

            val collection = MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
            val imageUri = contentResolver.insert(collection, values)!!

            contentResolver.openOutputStream(imageUri).use { bitmap.compress(Bitmap.CompressFormat.PNG, 100, it) }

            values.clear()
            values.put(MediaStore.Images.Media.IS_PENDING, 0)
            contentResolver.update(imageUri, values, null, null)

            Snackbar.make(findViewById(R.id.nametxt), "Saved: Pictures/posidon walls/$name", Snackbar.LENGTH_SHORT).show()
        } else {
            val direct = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) getExternalFilesDir("Wallpapers") else File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).toString() + "/Wallpapers")
            if (direct != null && (direct.exists() || direct.mkdirs())) {
                val file = File(direct, name.replace(' ', '_') + ".png")
                if (file.exists()) Snackbar.make(findViewById(R.id.nametxt), "Already Saved: " + file.absolutePath, Snackbar.LENGTH_SHORT).show() else try {
                    val out = FileOutputStream(file)
                    bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
                    out.flush()
                    out.close()
                    Snackbar.make(findViewById(R.id.nametxt), "Saved: " + file.absolutePath, Snackbar.LENGTH_SHORT).show()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    @SuppressLint("MissingSuperCall")
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == 0 && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) saveBitmap(img!!, Gallery.walls[index].name!!)
    }

    companion object {
        var img: Bitmap? = null
    }
}